### ISSUE TITLE
add `(a -> IO b) -> a -> Benchmarkable` benchmark function for IO?

### ISSUE THREAD

**Issue by rubenpieters** [bos/criterion#198]
I've been quite confused by the results of `nfIO`, since it seems that some evaluations are only measured once in the whole benchmark session. This causes the reported runtime to be very different than measuring with for example the `time` command. (See the reddit discussion [here](https://www.reddit.com/r/haskell/comments/8u0bf3/why_does_replicatem_2_not_add_runtime_to_this/) )

I made a `nfIOf :: (a -> IO b) -> a -> Benchmarkable` function which works similarly to `nf`/`whnf` (in a [fork](https://github.com/rubenpieters/criterion/blob/d018a693096efe69fac9d5f217efdeb4b5bac47a/Criterion/Types.hs#L368) ). The results of using this benchmarking function corresponds much more closely to using the `time` command.

I have set up a [repo](https://github.com/rubenpieters/conduit-bench/blob/bdd2f9e9430cb47314236b68492452238a528989/src/BenchCriterion.hs) where this behaviour can be observed.

In the log below, we can see that the results of using `nfIOf` correspond better to the timing with `time` than using `nfIO`.

```
> (/usr/bin/time -f %U stack exec bench >/dev/null)
5.63
> stack exec bench-criterion -- --csv bench.csv >/dev/null
> cat bench.csv
Name,Mean,MeanLB,MeanUB,Stddev,StddevLB,StddevUB
nfIOf,5.734192717041727,5.162777017875214,6.557142470999679,0.8074719138965863,0.2132544295833818,1.0733830262801656
nfIO,4.0720115103941275e-3,3.930939314357558e-3,4.26380839961976e-3,5.236585624969149e-4,4.0895005084673713e-4,7.094929138449197e-4
```

The relevant information being:

- using time - 5.63 s
- using nfIOf - 5.73 s
- using nfIO - 4.07e-3 s

**Comment by RyanGlScott** [comment]
As Syrak says on that reddit discussion, it's likely that `runPrimes 10000` fuses away to something quite efficient, so benchmarking it would be like running a thunk that's already been forced. Moreover, the documentation for `nfIO` doesn't really advertise this pitfall, which is ought to.

I think I'd be on-board with adding your proposed function of type `NFData b => (a -> IO b) -> a -> Benchmarkable` (as well as a version that evaluates only to WHNF). I'm not 100% sure about what to name these—perhaps `nfAppIO` and `whnfAppIO`?

**Comment by rubenpieters** [comment]
I created a pull request with the functions and your suggested naming. I added some small comments in the documentation for the functions, but I'm not sure how strong the warning for this scenario should be.

**Comment by RyanGlScott** [comment]
Fixed in #199.
